File: app/src/main/java/com/rakuishi/ok/ui/RepoAdapter.java
package com.rakuishi.ok.ui;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import com.rakuishi.ok.R;
import com.rakuishi.ok.api.model.Repo;

import java.util.List;

import butterknife.ButterKnife;
import butterknife.InjectView;

/**
 * Created by rakuishi on 15/05/03.
 */
public class RepoAdapter extends ArrayAdapter<Repo> {

    private LayoutInflater mLayoutInflater;

    public RepoAdapter(Context context, List<Repo> list) {
        super(context, 0, list);
        mLayoutInflater = LayoutInflater.from(context);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            convertView = mLayoutInflater.inflate(R.layout.row_three_line, parent, false);
            holder = new ViewHolder(convertView);
            convertView.setTag(holder);
        } else {
            holder = (ViewHolder)convertView.getTag();
        }

        Repo item = getItem(position);
        holder.primaryTextView.setText(item.getName());
        holder.secondaryTextView.setText(item.getDescription());
        holder.tertiaryTextView.setText(item.getUpdatedAt());

        return convertView;
    }

    class ViewHolder {
        @InjectView(R.id.row_three_line_primary_tv) TextView primaryTextView;
        @InjectView(R.id.row_three_line_secondary_tv) TextView secondaryTextView;
        @InjectView(R.id.row_three_line_tertiary_tv) TextView tertiaryTextView;

        public ViewHolder(View view) {
            ButterKnife.inject(this, view);
        }
    }
}

File: app/src/main/java/com/rakuishi/ok/presentation/repo/RepoAdapter.kt
package com.rakuishi.ok.presentation.repo

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.rakuishi.ok.R
import com.rakuishi.ok.data.Repo
import kotlinx.android.synthetic.main.list_item_repo.view.*

class RepoAdapter : ListAdapter<Repo, RecyclerView.ViewHolder>(RepoDiffCallback()) {

    var onItemClick: ((Repo) -> Unit)? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return RepoViewHolder(parent)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        (holder as RepoViewHolder).bind(getItem(position)) {
            onItemClick?.invoke(it)
        }
    }
}

private class RepoViewHolder(parent: ViewGroup) :
    RecyclerView.ViewHolder(
        LayoutInflater.from(parent.context).inflate(R.layout.list_item_repo, parent, false)
    ) {

    fun bind(repo: Repo, onItemClick: ((Repo) -> Unit)) {
        itemView.primaryTextView.text = repo.name
        itemView.secondaryTextView.text = repo.description
        itemView.backgroundLayout.setOnClickListener { onItemClick.invoke(repo) }
    }
}

private class RepoDiffCallback : DiffUtil.ItemCallback<Repo>() {
    override fun areItemsTheSame(oldItem: Repo, newItem: Repo): Boolean {
        return oldItem.id == newItem.id
    }

    override fun areContentsTheSame(oldItem: Repo, newItem: Repo): Boolean {
        return oldItem == newItem
    }
}